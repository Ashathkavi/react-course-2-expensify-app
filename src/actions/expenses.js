import uuid from 'uuid'
import database from '../firebase/firebase'


//ADD_EXPENSE
export const addExpense = (expense) => ({
    type:'ADD_EXPENSE',
    expense
})

export const startAddExpense = (expenseData ={}) => {
    return (dispatch)=>{
        const { description = '', note= '', amount=0, createdAt=0 } = expenseData
        const expense = {description , note, amount, createdAt}
        return database.ref('expenses').push(expense)
            .then((ref)=>{
                dispatch(addExpense({
                    id:ref.key,
                    ...expense
                }))
            })
    }
}
 

//REMOVE_EXPENSES
export const removeExpense = ({id} = {}) => ({
    type:'REMOVE_EXPENSE',
    id
})


//ASYNC REMOVE_EXPENSES

export const startRemoveExpenses = ({id}={}) => {
    return (dispatch)=>{
        //const { description = '', note= '', amount=0, createdAt=0 } = expenseData
        //const expense = {description , note, amount, createdAt}
        return database.ref(`expenses/${id}`).remove()
            .then(()=>{
                dispatch(removeExpense({id}))
            }).catch((error)=>{
                console.log("Remove failed: " + error.message)
            })                     
    }
}



//EDIT_EXPENSES
export const editExpense = (id, updates) => ({
    type:'EDIT_EXPENSE',
    id,
    updates
})

export const startEditExpenses = (id, updates) => {
    return (dispatch)=>{
        return database.ref(`expenses/${id}`).update({
            ...updates
        }).then(()=>{
                dispatch(editExpense(id, updates))
        }).catch((error)=>{
            console.log("Edit failed: " + error.message)
        })                     
    }
}



//SET_EXPENSES
export const setExpenses = (expenses) => ({
    type:'SET_EXPENSES',
    expenses
})


//ASYNC SET_EXPENSES

export const startSetExpenses = () => {
    return (dispatch)=>{
        //const { description = '', note= '', amount=0, createdAt=0 } = expenseData
        //const expense = {description , note, amount, createdAt}
        return database.ref('expenses')
            .once('value').then((snapshot)=>{
                const expenses = []
                snapshot.forEach((childSnapshot)=>{
                    expenses.push({
                        id:childSnapshot.key,
                        ...childSnapshot.val()
                    })
                })
                dispatch(setExpenses(expenses))
            }, (e)=>{
                console.log('Error with data fetching: ', e)
            })            
    }
}