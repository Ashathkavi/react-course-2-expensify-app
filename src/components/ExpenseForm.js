import React from 'react'
import moment from 'moment'
import DatePicker from 'react-datepicker'


const now = moment()
//console.log(now.format('MMM do, YYYY'))

class ExpenseForm extends React.Component{
    constructor(props){
        super(props)
        this.state= {
            description:props.expense ? props.expense.description : '', 
            note:props.expense ? props.expense.note : '', 
            amount:props.expense ? (props.expense.amount / 100).toString() : '', 
            createdAt:props.expense ?  moment(props.expense.createdAt).toDate() : moment().toDate(),
            error:''
        }
        
    }
    
    
    onDescriptionChange = (e) => { 
        e.persist()
        this.setState(()=>({description:e.target.value}))
    }
    onNoteChange = (e) => {
        e.persist()
        this.setState(()=>({note:e.target.value}))
    }
    onAmountChange = (e) => {
        const amount = e.target.value
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/gm)){
            this.setState(()=>({amount}))
        }        
    }
    onDateChange = (createdAt) => {
        if(createdAt){
            this.setState(() => ({createdAt}))
        }
    }
    onFocusChange = ({focused}) => {
        this.setState(()=>({calenderFocused:focused}))
    }
    onSubmit = (e) => {
        e.preventDefault()
        if(!this.state.description || !this.state.amount){
            this.setState(()=>({error:'Please provide description and amount.'}))
        }else{
            this.setState(()=>({error:''}))
            this.props.onSubmit({
                description:this.state.description,
                amount:parseFloat(this.state.amount, 10) * 100,
                createdAt:this.state.createdAt.valueOf(),
                note:this.state.note
            })
        }
    }
    render(){
        //console.log(this.state.description)
        return(
            <div>
            {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit = {this.onSubmit}>
                    <input 
                        typr="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}                        
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}                        

                    />
                    <DatePicker
                        selected = {this.state.createdAt}
                        onChange = {this.onDateChange}
                        closeOnScroll={e => e.target === document}
                    />
                    <textarea 
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>Add Expense</button>

                </form>
            </div>
        )
    }
}

export default ExpenseForm