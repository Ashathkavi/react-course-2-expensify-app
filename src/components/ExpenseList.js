import React from 'react'
import {connect} from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'


//exported for the testing purporse
export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No Expenses</p>
            ):(
                props.expenses.map((expense)=>(
                    <ExpenseListItem key={expense.id} {...expense}/>
                ))
            )
            
        }

    </div>
)


const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filter)
    }
}


const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList)

export default ConnectedExpenseList