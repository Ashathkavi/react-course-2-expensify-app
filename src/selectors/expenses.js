
import moment from 'moment'


//get visible expenses
const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    //console.log(startDate, "startDate")
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMoment = moment(startDate)
        const endDateMoment = moment(endDate)
        //console.log(startDateMoment,'startDateMoment')
        // console.log(createdAtMoment,'createdAtMoment')
        // console.log(endDateMoment,'endDateMoment')

        const startDateMatch = startDate ? startDateMoment.isSameOrBefore(createdAtMoment, 'day'):true
        //const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        const endDateMatch = endDate ? endDateMoment.isSameOrAfter(createdAtMoment, 'day'):true
        //const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())

        return startDateMatch && endDateMatch && textMatch

    }).sort((a,b)=>{
        if(sortBy === 'date'){
            return a.createdAt < b.createdAt ? 1 : -1
        }else if(sortBy === 'amount'){
            return a.amount < b.amount ? 1 : -1
        }
    })
}

export default getVisibleExpenses