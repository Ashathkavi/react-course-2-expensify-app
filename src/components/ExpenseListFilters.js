import React from 'react'
import {connect} from 'react-redux'
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters'
import DatePicker from 'react-datepicker'
import moment from 'moment'

export class ExpenseListFilters extends React.Component{

    onDateChange = (dates) => {
        let [startDate, endDate] = dates;
        startDate = moment(startDate).valueOf()
        endDate =  moment(endDate).valueOf()
        //console.log(startDate,'start',endDate,'end')
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }

    onSortChange = (e) => {
        e.target.value === "date" ? 
            this.props.sortByDate() 
        : 
            this.props.sortByAmount()
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }

    render(){
        return(
            <div>
                <input type="text" value={this.props.filters.text} onChange={this.onTextChange}/>
                <select 
                    value={this.props.filters.sortBy} 
                    onChange={this.onSortChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DatePicker
                    selected={this.props.filters.startDate}
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onChange= {this.onDateChange}
                    selectsRange
                    inline
                />

            </div>
        )
    }
}


const mapStateToProps = (state) => {
    return{
        filters:state.filter
    }
} 

const mapDispatchToProps = (dispatch) => ({
    setTextFilter:(text) => dispatch(setTextFilter(text)), 
    sortByAmount:() => dispatch(sortByAmount()), 
    sortByDate:() => dispatch(sortByDate()), 
    setStartDate:(startDate) => dispatch(setStartDate(startDate)), 
    setEndDate:(enDate) => dispatch(setEndDate(enDate))
})

const connectedExpenseListFilters = connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)

export default connectedExpenseListFilters