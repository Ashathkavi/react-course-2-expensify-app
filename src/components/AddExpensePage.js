import React from 'react'
import ExpenseForm from './ExpenseForm'
import {connect} from 'react-redux'
import {addExpense} from '../actions/expenses'

export class AddExpensePage extends React.Component{
    onSubmit=(expense) => {
        //props.dispatch(addExpense(expense))
        this.props.addExpense(expense)
        this.props.history.push('/')
    }
    render(){
        return(
            <div>
                This is from AddExpensePage component
                <ExpenseForm onSubmit={this.onSubmit}  />
            </div>
        )
    }
}


// this function makes the AddExpensePage component testable
const mapDispatchToProps = (dispatch) => ({
    addExpense:(expense) =>dispatch(addExpense(expense))
})

export default connect(undefined,mapDispatchToProps)(AddExpensePage)