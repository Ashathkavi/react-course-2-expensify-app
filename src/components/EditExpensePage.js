import React from 'react'
import ExpenseForm from './ExpenseForm'
import {connect} from 'react-redux'
import {editExpense, startEditExpenses} from '../actions/expenses'
import {startRemoveExpenses} from '../actions/expenses'

export class EditExpensePage extends React.Component{
    

    onSubmit = (expense) => {
        this.props.editExpense(this.props.match.params.id,expense)
        this.props.history.push('/')
    }

    onRemove = () => {
        this.props.removeExpense({id:this.props.expense.id});
        this.props.history.push('/')
    }

    render(){
        return(
            <div>
                This is from EditExpensePage component
                <button onClick={this.onRemove}>Remove</button>
                <ExpenseForm onSubmit={this.onSubmit} expense ={this.props.expense}/>
            </div>
        )
    }
}


const mapStateToProps = (state, props) => {
    return{
        expense:state.expenses.find((expense)=>{
            //console.log(expense)
            return expense.id === props.match.params.id
        })
    }
}

const mapDispatchToProps = (dispatch, props) => ({
    editExpense:(id,expense) =>dispatch(startEditExpenses(id,expense)),
    removeExpense:(date) => dispatch(startRemoveExpenses(date))
})

export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage) 