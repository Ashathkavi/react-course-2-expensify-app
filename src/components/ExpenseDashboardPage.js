import React from 'react'
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import ExpencesSummary from './ExpenseSummary'

const ExpenseDashboardPage = () => (
    <div>
        <ExpencesSummary/>
        <ExpenseListFilters/>
        <ExpenseList/>
    </div>
)

export default ExpenseDashboardPage