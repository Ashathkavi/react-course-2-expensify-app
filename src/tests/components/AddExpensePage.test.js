import React from 'react'
import {shallow} from 'enzyme'
import {AddExpensePage} from '../../components/AddExpensePage'
import expenses from '../fixtures/expenses'
import ExpenseForm from '../../components/ExpenseForm'

let addExpenseSpy, historySpy, wrapper;
beforeEach(()=>{
    addExpenseSpy = jest.fn()
    historySpy = {push:jest.fn()}
    wrapper = shallow(<AddExpensePage history={historySpy} addExpense={addExpenseSpy}/>)
})

test('should render AddExpensePage correctly', ()=>{
    expect(wrapper).toMatchSnapshot()
})


test('should handle onSubmit', ()=>{
    wrapper.find(ExpenseForm).prop('onSubmit')(expenses[0])
    expect(historySpy.push).toHaveBeenLastCalledWith('/')
    expect(addExpenseSpy).toHaveBeenLastCalledWith(expenses[0])
    
})