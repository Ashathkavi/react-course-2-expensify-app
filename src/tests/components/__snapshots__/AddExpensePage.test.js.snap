// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddExpensePage correctly 1`] = `
<div>
  This is from AddExpensePage component
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
